### PR TITLE
Wicket 6786

### PR DESCRIPTION
**Pull request by eozmen410**
Initial implementation of fetch metadata header checks in CsrfPreventionRequestCycleListener and the related tests

### CODE REVIEW

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java [hunk 1]
```diff
@@ -312,6 +314,15 @@ protected boolean isChecked(IRequestHandler handler)
 			!(handler instanceof RenderPageRequestHandler);
 	}
 
+	/**
+	 * Checks if Sec-Fetch-* headers are present
+	 */
+	protected boolean hasFetchMetadataHeaders(HttpServletRequest containerRequest)
+	{
+		String secFetchSiteValue = containerRequest.getHeader(ResourceIsolationPolicy.SEC_FETCH_SITE_HEADER);
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java]: let's add static imports for fields under the ResourceIsolationPolicy interface, please

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java [hunk 2]
```diff
@@ -345,13 +356,21 @@ public void onRequestHandlerResolved(RequestCycle cycle, IRequestHandler handler
 			IRequestablePage targetedPage = prh.getPage();
 			HttpServletRequest containerRequest = (HttpServletRequest)cycle.getRequest()
 				.getContainerRequest();
-			String sourceUri = getSourceUri(containerRequest);
+            String sourceUri = getSourceUri(containerRequest);
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java]: let's keep indentation throughout our commits. You can use the **Reformat code** option on IntelliJ to take care of this automatically.

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java [hunk 3]
```diff
@@ -451,6 +470,54 @@ protected void checkRequest(HttpServletRequest request, String sourceUri, IReque
...
+     * @param sourceUri
+     *          the source URI
+     * @param page
+     *          the page that is the target of the request
+	 */
+	protected void checkRequestFetchMetadata(HttpServletRequest request, String sourceUri, IRequestablePage page)
+    {
+		//check if sourceUri exists before checking for whitelisted hosts,
+		// if not set sourceUri to no origin for logging purposes
+    	if (sourceUri == null || sourceUri.isEmpty())
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java]: let's turn this guard into a single check

```suggestion
    	if (Strings.isEmpty(sourceUri))
```

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java [hunk 4]
```diff
@@ -451,6 +470,54 @@ protected void checkRequest(HttpServletRequest request, String sourceUri, IReque
...
+                case ABORT :
+                    abortHandler(request, sourceUri, page);
+                    break;
+            }
+        }
+        else
+        {
+            matchingOrigin(request, sourceUri, page);
+        }
+    }
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java]: reducing nestedness by doing the positive check first makes this code slightly easier to read:

```suggestion
        if (fetchMetadataPolicy.isRequestAllowed(request))
        {
            matchingOrigin(request, sourceUri, page);
            return;
        }
        
        log.debug("Request is not allowed by the resource isolation policy, {}", conflictingOriginAction);
        switch (conflictingOriginAction)
        {
            case ALLOW :
                allowHandler(request, sourceUri, page);
                break;
            case SUPPRESS :
                suppressHandler(request, sourceUri, page);
                break;
            case ABORT :
                abortHandler(request, sourceUri, page);
                break;
        }
    }
```

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package org.apache.wicket.protocol.http;
+
+import org.apache.wicket.protocol.http.ResourceIsolationPolicy;
+
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java]: let's apply **Reformat code** to this file to remove this white space and fix formatting generally

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import javax.servlet.http.HttpServletRequest;
+
+public class DefaultResourceIsolationPolicy implements ResourceIsolationPolicy {
+
+  @Override
+  public boolean isRequestAllowed(HttpServletRequest request) {
+    String site = request.getHeader(SEC_FETCH_SITE_HEADER);
+
+    // TODO: Is this needed?
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java]: According to [the Fetch Metadata spec on this section header](https://www.w3.org/TR/fetch-metadata/#sec-fetch-site-header) the value should be lower case. Checking [Chrome's source code](https://source.chromium.org/chromium/chromium/src/+/master:services/network/sec_header_helpers.cc;l=39;bpv=0;bpt=1?originalUrl=https:%2F%2Fcs.chromium.org%2F) confirms that this value should always be lower case. I would be tempted to skip this step!

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    if (SAME_ORIGIN.equals(site) || SAME_SITE.equals(site) || NONE.equals(site)) {
+      return true;
+    }
+
+    // Allow simple top-level navigations except <object> and <embed>
+    return isAllowedTopLevelNavigation(request);
+  }
+
+  private boolean isAllowedTopLevelNavigation(HttpServletRequest request) {
+    // TODO: Assert these headers are never null at this point?
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java]: I couldn't find any assertions in the existing code. Perhaps we could skip the assertions this is referring to.

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/ResourceIsolationPolicy.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  String SEC_FETCH_SITE_HEADER = "sec-fetch-site";
+  String SEC_FETCH_MODE_HEADER = "sec-fetch-mode";
+  String SEC_FETCH_DEST_HEADER = "sec-fetch-dest";
+  String SAME_ORIGIN = "same-origin";
+  String SAME_SITE = "same-site";
+  String NONE = "none";
+  String MODE_NAVIGATE = "navigate";
+  String DEST_OBJECT = "object";
+  String DEST_EMBED = "embed";
+  // TODO: Add remaining Fetch Metadata keywords
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/ResourceIsolationPolicy.java]: do we have all headers on the spec now? can we remove this TODO?

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/ResourceIsolationPolicy.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
+package org.apache.wicket.protocol.http;
+
+import javax.servlet.http.HttpServletRequest;
+
+@FunctionalInterface
+public interface ResourceIsolationPolicy {
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/ResourceIsolationPolicy.java]: let's add a Javadoc on this class giving a small explanation of what resource isolation policies are and pointing to the spec, please. Let's add an @author annotation too!

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java [hunk 10]
```diff
@@ -451,6 +470,54 @@ protected void checkRequest(HttpServletRequest request, String sourceUri, IReque
...
+		{
+			// if the origin is a know and trusted origin, don't check any further but allow the request
+			if (isWhitelistedHost(sourceUri)) {
+				whitelistedHandler(request, sourceUri, page);
+				return;
+			}
+		}
+        if (!fetchMetadataPolicy.isRequestAllowed(request))
+        {
+            log.debug("Request is not allowed by the resource isolation policy, {}", conflictingOriginAction);
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java]: logging the URL in this statement could be useful to investigate breakages and potential attacks

**Review by salcho — CHANGES_REQUESTED** [review]
Nice work Jen! Only a few formatting issues but the logic looks solid. Let's address these comments for now and wait for Aaron's review next week :)

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
+package org.apache.wicket.protocol.http;
+
+import org.apache.wicket.protocol.http.ResourceIsolationPolicy;
+
+import javax.servlet.http.HttpServletRequest;
+
+public class DefaultResourceIsolationPolicy implements ResourceIsolationPolicy {
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java]: let's add a small Javadoc on this class pointing to https://web.dev/fetch-metadata/ (which is the inspiration for this default policy) and let's add ourselves as authors using the author annotation :)

**Review by salcho — CHANGES_REQUESTED** [review]

**Comment by salcho** [comment]
this change is still missing the Vary header changes, which Jen is currently working on

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java [hunk 12]
```diff
@@ -451,6 +476,54 @@ protected void checkRequest(HttpServletRequest request, String sourceUri, IReque
...
+		}
+		else
+		{
+			// if the origin is a know and trusted origin, don't check any further but allow the request
+			if (isWhitelistedHost(sourceUri))
+			{
+				whitelistedHandler(request, sourceUri, page);
+				return;
+			}
+		}
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java]: fix indentation on this block, please

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * Default resource isolation policy used in {@link CsrfPreventionRequestCycleListener}
+ * that implements the {@link ResourceIsolationPolicy} interface. This default
+ * policy is based on <a href="https://web.dev/fetch-metadata/">https://web.dev/fetch-metadata/</a>.
+ *
+ * @see <a href="https://web.dev/fetch-metadata/">https://web.dev/fetch-metadata/</a>
+ *
+ * @author Santiago Diaz
+ * @author Ecenaz Jen Ozmen
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java]:
```suggestion
 * @author Santiago Diaz - saldiaz@google.com
 * @author Ecenaz Jen Ozmen - ecenazo@google.com
```

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/ResourceIsolationPolicy.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+ * origin attacks and use the {@code sec-fetch-*} request headers to decide
+ * whether to accept or reject a request. Read more about
+ * <a href="https://web.dev/fetch-metadata/">Fetch Metadata.</a>
+ *
+ * See {@link DefaultResourceIsolationPolicy} for the default implementation used.
+ *
+ * @see <a href="https://web.dev/fetch-metadata/">https://web.dev/fetch-metadata/</a>
+ *
+ * @author Santiago Diaz
+ * @author Ecenaz Jen Ozmen
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/ResourceIsolationPolicy.java]:
```suggestion
 * @author Santiago Diaz - saldiaz@google.com
 * @author Ecenaz Jen Ozmen - ecenazo@google.com
```

**Review by salcho — CHANGES_REQUESTED** [review]

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java [hunk 15]
```diff
@@ -345,13 +359,24 @@ public void onRequestHandlerResolved(RequestCycle cycle, IRequestHandler handler
...
 			if (isChecked(targetedPage))
 			{
-				// if so check the Origin HTTP header
-				checkRequest(containerRequest, sourceUri, targetedPage);
+				//check sec-fetch-site header and call the fetch metadata check if present
+				if (hasFetchMetadataHeaders(containerRequest))
+				{
+					WebResponse webResponse = (WebResponse)cycle.getResponse();
+					//TODO: should we define "Vary" as a header in ResourceIsolationPolicy / somewhere else?
+					webResponse.setHeader("Vary", "Sec-Fetch-Dest, Sec-Fetch-Site, Sect-Fetch-Mode");
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java]: there's a typo in Sec-Fetch-Mode here. Let's write a test that verifies that these headers are set to the correct value and let's use constants like we do in the default resource isolation policy, please.

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+        return isAllowedTopLevelNavigation(request);
+    }
+
+    private boolean isAllowedTopLevelNavigation(HttpServletRequest request)
+    {
+        String mode = request.getHeader(SEC_FETCH_MODE_HEADER);
+        String dest = request.getHeader(SEC_FETCH_DEST_HEADER);
+
+        // TODO: Does the method need to be upper cased?
+        boolean isSimpleTopLevelNavigation = MODE_NAVIGATE.equals(mode) || "GET".equals(request.getMethod());
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java]: let's use the javax.servlet.http.HttpServlet#METHOD_GET field here and remove this TODO, please

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java (hunk 15 above)
Comment (salcho) [reply wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java]: +1 on defining the Vary keyword as a constant associated with ResourceIsolationPolicy

**Review by salcho — CHANGES_REQUESTED** [review]

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java (hunk 16 above)
Comment (eozmen410) [reply wicket-core/src/main/java/org/apache/wicket/protocol/http/DefaultResourceIsolationPolicy.java]: it seems like METHOD_GET is a private variable: http://api.dpml.net/javax/servlet/2.5/constant-values.html#javax.servlet.http.HttpServlet.METHOD_GET, IntelliJ doesn't recognize it when I try to do `HttpServlet.METHOD_GET`

File: wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java [hunk 17]
```diff
@@ -350,15 +365,29 @@ public void onRequestHandlerResolved(RequestCycle cycle, IRequestHandler handler
...
-				checkRequest(containerRequest, sourceUri, targetedPage);
+				// check sec-fetch-site header and call the fetch metadata check if present
+				if (hasFetchMetadataHeaders(containerRequest))
+				{
+					// set vary headers to avoid caching responses processed by Fetch Metadata
+					// caching these responses may return 403 responses to legitimate requests
+					// or defeat the protection
+					WebResponse webResponse = (WebResponse)cycle.getResponse();
+					webResponse.setHeader(VARY_HEADER, SEC_FETCH_DEST_HEADER + ", "
+						+ SEC_FETCH_SITE_HEADER + ", " + SEC_FETCH_MODE_HEADER);
```
Comment (salcho) [inline wicket-core/src/main/java/org/apache/wicket/protocol/http/CsrfPreventionRequestCycleListener.java]: We should try to be as little intrusive as possible. Let's only set the Vary header if we act upon the current request, so let's move this inside checkRequestFetchMetadata, after checking if this host is in our allowlist (only then we will make a judgement based on FM)

**Review by salcho — CHANGES_REQUESTED** [review]
can we add tests for the Vary behaviour, please?

**Review by salcho — APPROVED** [review]

**Pull request merged**